include/lm/language_model.h

```diff
@@ -26,87 +51,74 @@ class language_model
      * Creates an N-gram language model based on the corpus specified in the
      * config file.
      */
-    language_model(const std::string& config_file);
-
-    /**
-     * Creates an N-gram language model based on the corpus specified in the
-     * config file.
-     * @param n The value of n, which overrides any setting in the config file
-     */
-    language_model(const std::string& config_file, size_t n);
+    language_model(const cpptoml::table& config);
 
     /**
-     * Randomly generates one token sequence based on <s> and </s> symbols.
-     * @return a random sequence of tokens based on this language model
+     * Default move constructor.
      */
-    std::string generate(unsigned int seed) const;
+    language_model(language_model&&) = default;
 
     /**
-     * @param tokens The previous N - 1 tokens
-     * @param random A random number on [0, 1] used for choosing the next token
-     * @return the next token based on the previous tokens
-     */
-    std::string next_token(const std::deque<std::string>& tokens,
-                           double random) const;
-
-    /**
-     * @param tokens A sequence of tokens
+     * @param sentence A sequence of tokens
      * @return the perplexity of this token sequence given the current language
      * model: \f$ \sqrt[n]{\prod_{i=1}^n\frac{1}{p(w_i|w_{i-n}\cdots w_{i-1})}}
      * \f$
      */
-    double perplexity(const std::string& tokens) const;
+    float perplexity(const sentence& tokens) const;
 
     /**
-     * @param tokens A sequence of tokens
+     * @param sentence A sequence of tokens
      * @return the perplexity of this token sequence given the current language
      * model normalized by the length of the sequence
      */
-    double perplexity_per_word(const std::string& tokens) const;
+    float perplexity_per_word(const sentence& tokens) const;
 
     /**
-     * @param tokens A sequence of n tokens
-     * @return the probability of seeing the nth token based on the previous n
-     * - 1 tokens
+     * @param tokens A sequence of n tokens (one sentence)
+     * @return the log probability of the likelihood of this sentence
      */
-    double prob(std::deque<std::string> tokens) const;
+    float log_prob(sentence tokens) const;
 
-  private:
+    /**
+     * @param prev Seen tokens to base the next token off of
+     * @param k Number of results to return
+     * @return a sorted vector of likely next tokens
+     */
+    std::vector<std::pair<std::string, float>> top_k(const sentence& prev,
+                                                     size_t k) const;
```

Comment:
Again it might be nice to note complexity guarantees here (since I assume this is inefficient right now?).